### PR TITLE
Update bf redesign

### PR DESCRIPTION
**Pull request by tibor95**
Reworking update_buildable_field():
- breaking into multiple functions to make code smaller/easier to read
- replacing find_immovables() with map area walking

### CODE REVIEW

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc]:
```suggestion
	Widelands::HollowArea<> har(
```

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc]:
```suggestion
	Widelands::MapHollowRegion<> second_area(map, har);
```

**Comment by tibor95** [comment]
I accepted your proposals, we will see

**Comment by tibor95** [comment]
Build failed, I had these error in my PC too:
https://github.com/widelands/widelands/runs/4948048221?check_suite_focus=true#step:5:12

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc]:
```suggestion
	// TODO(tibor): I was not able to define that hollow area :(
```

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc]:
```suggestion
	   Widelands::Area<>(map.get_fcoords(field.coords), kProductionArea + 2),
```

**Comment by hessenfarmer** [comment]
ok found it maphollowregion does only provide the coords of a position rather then Fcoords so we first need to fetch the fcoords before we are doing stuf
[fcoords_patch.txt](https://github.com/widelands/widelands/files/7941558/fcoords_patch.txt)
f

**Comment by tibor95** [comment]
Finally green!

are function names like 'handle_*_site' fine?

**Comment by hessenfarmer** [comment]
I am really curious whether this gained us some performance. do you have any comparison results? my machine is too fast to provide meaningful results.

**Comment by hessenfarmer** [comment]
> Finally green!
> 
> are function names like 'handle_*_site' fine?

basically yes. However the pure name might be misleading if put out of context (handle might mean control the site) so perhaps consider_* might be more clear. your decision.   

please convert this branch from draft to real so we can start a final review and get this in.

**Comment by tibor95** [comment]
> I am really curious whether this gained us some performance. do you have any comparison results? my machine is too fast to provide meaningful results.

No. But from where the performance gain would come? The code is doing little bit more, more function calls.

I can do some performance test, but unless the difference will be big I do not dare to draw a conclusion

**Comment by hessenfarmer** [comment]
> > I am really curious whether this gained us some performance. do you have any comparison results? my machine is too fast to provide meaningful results.
> 
> No. But from where the performance gain would come? The code is doing little bit more, more function calls.    
  
because we are not checking the inner fields twice and we are walking directly over the fields now and not first walk them and afterwards iterate over the results. 

> 
> I can do some performance test, but unless the difference will be big I do not dare to draw a conclusion     
   
I'd really love to see the results. thanks a lot

**Comment by tibor95** [comment]
```
master (debug)

[04:01:39.298 game] DEBUG: Player: 4, AI tasks statistics:  call count  ms total   avg     max
[04:01:39.298 game] DEBUG:   check buildable fields      :     298      11425      38      60
[04:01:39.298 game] DEBUG:   print statistics            :       5          7       1       3
[04:01:39.298 game] DEBUG:   check militarysites         :     239        455       2       4
[04:01:39.298 game] DEBUG:   roads check                 :     680       1619       2      23
[04:01:39.298 game] DEBUG:   check ships                 :     295          1       0       0
[04:01:39.298 game] DEBUG:   construct a building        :     281       1890       7      15
[04:01:39.298 game] DEBUG:   check mines                 :     137         12       0       0
[04:01:39.298 game] DEBUG:   check economies             :     215          8       0       0
[04:01:39.298 game] DEBUG:   check unbuildable fields    :     338       2216       7      37
[04:01:39.298 game] DEBUG:   check mineable fields       :     107        133       1      13
[04:01:39.298 game] DEBUG:   check enemy sites           :     107        193       2       4
[04:01:39.298 game] DEBUG:   productionsites statistics  :     182         17       0       0
[04:01:39.298 game] DEBUG:   check training sites        :     131          3       0       0
[04:01:39.298 game] DEBUG:   flag warehouse Update       :     131        317       2       7
[04:01:39.298 game] DEBUG:   productionsites check       :     136         63       0       3
[04:01:39.298 game] DEBUG:   marine decisions            :     104          5       0       0
[04:01:39.298 game] DEBUG:   count military vacant       :      84         12       0       0
[04:01:39.298 game] DEBUG:   update player stats         :      39          0       0       0
[04:01:39.298 game] DEBUG:   wares review                :       3          0       0       0




this branch (debug)

[04:01:34.914 game] DEBUG: Player: 4, AI tasks statistics:  call count  ms total   avg     max
[04:01:34.914 game] DEBUG:   check buildable fields      :     300       9248      31      47
[04:01:34.914 game] DEBUG:   check unbuildable fields    :     343       1895       6      32
[04:01:34.914 game] DEBUG:   print statistics            :       5          6       1       2
[04:01:34.914 game] DEBUG:   update player stats         :      38          0       0       0
[04:01:34.914 game] DEBUG:   marine decisions            :     102          5       0       0
[04:01:34.914 game] DEBUG:   roads check                 :     678       1858       3      22
[04:01:34.914 game] DEBUG:   check economies             :     212         10       0       0
[04:01:34.914 game] DEBUG:   check ships                 :     297          1       0       0
[04:01:34.914 game] DEBUG:   productionsites check       :     145         52       0       2
[04:01:34.914 game] DEBUG:   flag warehouse Update       :     132        320       2       8
[04:01:34.914 game] DEBUG:   construct a building        :     272       1964       7      22
[04:01:34.914 game] DEBUG:   check mines                 :     137         13       0       0
[04:01:34.914 game] DEBUG:   check mineable fields       :     108        119       1       2
[04:01:34.914 game] DEBUG:   check training sites        :     132          3       0       0
[04:01:34.914 game] DEBUG:   productionsites statistics  :     182         17       0       0
[04:01:34.914 game] DEBUG:   check militarysites         :     236        386       2      18
[04:01:34.914 game] DEBUG:   check enemy sites           :     107        200       2       4
[04:01:34.914 game] DEBUG:   count military vacant       :      86         13       0       0
[04:01:34.914 game] DEBUG:   wares review                :       3          0       0       0

```

I am afraid that master was compiled as a DEBUG, Visual Studio Code said something like that, but it was confusing. Anyway here you are. The results look quite different

**Comment by hessenfarmer** [comment]
Well it seems we gained a lot of performance by this change as I was hoping. However we should be sure to compare debug with debug and release with release builds. 
Anyhow I think this branch is ready for proper review so I convert to "ready for review" and merge master. After that I'll do a complete review.

**Comment by hessenfarmer** [comment]
ok did some testing with debug builds 
average time on master is 17-21 after 2 hours.
in this branch it is 12- 15.
it is not that big but still some 30% performance gain, so it was worth the effort.

**Comment by tibor95** [comment]
I updated my statistics above, both are now for DEBUG builds. My statistics shows about 20% improvement only, but I tested on savegame with many buildable fields, so the % can be dependent on the count of fields.

**Comment by hessenfarmer** [comment]
> I updated my statistics above, both are now for DEBUG builds. My statistics shows about 20% improvement only, but I tested on savegame with many buildable fields, so the % can be dependent on the count of fields.

definitly. I tested with 8 AI on Ice wars. And the variety of results was as big as the average performance gain however even a 20% performnce gain would be a v ery good result.

**Comment by hessenfarmer** [comment]
While activating my test i discovered that we have 2 definitions of kPrintStats: one is a bool and the other is a scheduler task id.
it would be nice to fix this in this branch
furthermore please disable stats printing and perf collection in this branch before merging as it is currently activated in master

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc]: we might skip this condition and just upcast as we already ensured befor the call to this function we have an immovable that is at least a building. Although it shouldn't hurt much to keep it

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc:1871]: we should move this check up to only have buildings in the unique_serials like in the sequence for the outer area

File: src/ai/ai_help_structs.h
Comment (hessenfarmer) [inline src/ai/ai_help_structs.h:362]: i don't understand could you please explain in short words how this works

**Comment by tibor95** [comment]
I slightly reworked the logic around kPrintStats, you will see

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc]: it seems this variable is just incremented but never used anywhere. so we could delete it

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc]: this variable is only used to determine field.unconnected_nearby which is not used anywhere after it has been set so it should be safe to delete both variables
see https://github.com/widelands/widelands/pull/5013/commits/1dfda555f8d26f8016cce459b2da0982e647be1e

**Review by hessenfarmer — APPROVED** [review]
some small nits inline 

rest looks very good thanks for this great work

File: src/ai/ai_help_structs.h
Comment (tibor95) [reply src/ai/ai_help_structs.h:362]: I was afraid when passing BF to those consider_* functions that they will be copy-constructed in that new functions and all info put into BF will be lost when functions exits (deleting the instance).
But this does not happen, so I can remove this "safe-guard".
I do not care much

File: src/ai/defaultai.cc
Comment (tibor95) [reply src/ai/defaultai.cc]: OK, removing for performance reasons :)

File: src/ai/defaultai.cc
Comment (tibor95) [reply src/ai/defaultai.cc]: yes, I was using it for debug. removed

File: src/ai/defaultai.cc
Comment (tibor95) [reply src/ai/defaultai.cc]: Good catch. However as we already have this info, I will try to use it in genetic algorithm. I put a todo there. Probably in my next branch

File: src/ai/defaultai.cc
Comment (tibor95) [reply src/ai/defaultai.cc:1871]: moved.

**Comment by hessenfarmer** [comment]
ok I will merge after the final build is green thanks for this great work

File: src/ai/defaultai.cc
Comment (hessenfarmer) [inline src/ai/defaultai.cc]: if we ever use this we might need to reconsider the condition here. but for now it is ok.

Comment (tibor95) [reply src/ai/defaultai.cc]: yes, I will review all this later.

**Pull request merged**
